Introduce an extra `TimelineItemContent` layer that holds aggregations

stefanceriu opened the pull request:
In preparation for threads we have realised that the `reactions`, `thread_root` and `in_reply_to` were only available on `Message` types, which doesn't play well with Stickers and Polls.

This PR introduces a new `Aggregated` `TimelineItemContent` variant which holds the message `kind` (Message, Sticker, Poll) as well as well as any related aggregated data. it will help treat them all in a similar fashion as well as account for future changes.

There are no functional changes, it's mostly about moving code around and the FFI interfaces haven't changed.

Best reviewed commit by commit.

bnjbvr commented on crates/matrix-sdk-ui/src/timeline/tests/encryption.rs:
```diff
@@ -855,10 +855,11 @@ async fn test_retry_decryption_updates_response() {
     // We receive the text response.
     {
         let event = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+        let aggregated = event.content().as_aggregated().unwrap();
         let msg = event.content().as_message().unwrap();
```
:warning:  Here and below (in pretty much all the tests this commit touches, unfortunately), it's a bit weird to get the as_aggregated() first, and then as_message(); this makes me think that `as_message()` should be a sub-method of `AggregatedTimelineItemContent`, and not directly live on `TimelineItemContent` (or should be a shortcut to go through the aggregated then the message, since it's so common to test for messages).

Anyhow, i think the test should first get the aggregated one, *then* get the message out of the aggregated one; otherwise it looks like getting the `aggregated` is spurious. Alternatively, we could have these tests use `AggregatedTimelineItemContent::in_reply_to()` directly, instead of storing the aggregated content in a local variable, then reading its `in_reply_to` internal value.

bnjbvr reviewed (approved):
Thanks, great work, and I can feel your pain having to go through all the test changes :people_hugging: 

I'm not a big fan of the name `AggregatedTimelineItemContent`, which is a mouthful. I would advocate for a shorter name; my reasoning is that `TImelineItem` is inferred from the surrounding context. And `Aggregated` isn't very informative unfortunately, so I'm tempted to steal Ruma's naming here, and call these `MessageLike`.

TL;DR, how about `MessageLikeContent` / `MessageLikeContentKind`?

I've got a few other non-trivial comments in the PR, but in the spirit of moving forward with this: approving. If some of the other comments feel too heavyweight to handle right now: please open follow-up issues, cc me/you on them, so we can take a look at those quickly. Thanks!

stefanceriu commented on crates/matrix-sdk-ui/src/timeline/event_item/content/mod.rs:
```diff
@@ -78,18 +78,42 @@ pub use self::{
+            AggregatedTimelineItemKind::Poll(_) => "a poll",
+        }
+    }
+}
```
It's gone now. I might or might not have messed things up during the rebase 🙈

stefanceriu commented on bindings/matrix-sdk-ffi/src/timeline/content.rs:
```diff
@@ -33,8 +33,9 @@ impl From<matrix_sdk_ui::timeline::TimelineItemContent> for TimelineItemContent
         match value {
-            Content::Aggregated(AggregatedTimelineItem {
-                kind: AggregatedTimelineItemKind::Message(message),
+            Content::Aggregated(AggregatedTimelineItemContent {
```
Oupsy, my bad, sorry about that 😢

stefanceriu commented on crates/matrix-sdk-ui/src/timeline/event_handler.rs:
```diff
@@ -580,21 +581,26 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
             .unzip();
 
-        let edit_json = edit_json.flatten();
+        let mut replied_to_event_id = None;
+        let mut thread_root = None;
+        let in_reply_to_details = msg.relates_to.clone().and_then(|relation| match relation {
```
Decided to leave this alone as `and_then` want to return an optional and clashes with the tuple.

stefanceriu commented on crates/matrix-sdk-ui/src/timeline/event_item/content/mod.rs:
```diff
@@ -310,13 +345,19 @@ impl TimelineItemContent {
+        let aggregated = as_variant!(self, Self::Aggregated)
+            .map(|f| as_variant!(&f.kind, AggregatedTimelineItemKind::Message));
+
+        aggregated.unwrap_or(None)
```
`as_variant` does not currently handle nested structs but that's being addressed in https://github.com/jplatte/as_variant/pull/1

stefanceriu commented on crates/matrix-sdk-ui/src/timeline/event_item/content/mod.rs:
```diff
@@ -500,6 +525,21 @@ impl TimelineItemContent {
+            | TimelineItemContent::FailedToParseState { .. }
+            | TimelineItemContent::CallInvite
+            | TimelineItemContent::CallNotify => None,
+        }
```
The inner values are themselves options so we decided on going with e.g.
`as_variant!(self, TimelineItemContent::Aggregated).and_then(|agg| agg.thread_root.clone())`

stefanceriu commented on crates/matrix-sdk-ui/src/timeline/event_item/content/mod.rs:
```diff
@@ -310,13 +345,19 @@ impl TimelineItemContent {
+        let aggregated = as_variant!(self, Self::Aggregated)
+            .map(|f| as_variant!(&f.kind, AggregatedTimelineItemKind::Poll));
+
+        aggregated.unwrap_or(None)
```
`as_variant` does not currently handle nested structs but that's being addressed in https://github.com/jplatte/as_variant/pull/1

stefanceriu commented:
Thank for you the great review! 🙏  
I've addressed most of the issues but in the interest of not losing any more time on conflicts I will merge this now and handle whatever's left unresolved (rename to `MessageLike`, cleanup `fetch_replied_to_event` and introduce more helpers to help with the tests) in a separate PR.
